sky/provision/gcp/instance_utils.py

```diff
@@ -1,21 +1,127 @@
+
+# This is the maximum number of times we will retry a GCP API call.
+# The number is identical to those we use for AWS boto3.
+GCP_MAX_RETRIES = 12
+GCP_CREATE_MAX_RETRIES = 5
+GCP_RETRY_INTERVAL_SECONDS = 5
+GCP_TIMEOUT = 300
 
-# Using v2 according to
-# https://cloud.google.com/tpu/docs/managing-tpus-tpu-vm#create-curl # pylint: disable=line-too-long
-TPU_VERSION = 'v2'
+logger = sky_logging.init_logger(__name__)
 
 _FIREWALL_RESOURCE_NOT_FOUND_PATTERN = re.compile(
     r'The resource \'projects/.*/global/firewalls/.*\' was not found')
 
 
+def _retry_on_http_exception(
+    regex: Optional[str] = None,
+    max_retries: int = GCP_MAX_RETRIES,
+    retry_interval_s: int = GCP_RETRY_INTERVAL_SECONDS,
+):
+    """Retry a function call n-times for as long as it throws an exception."""
+
+    def dec(func):
+
+        @functools.wraps(func)
+        def wrapper(*args, **kwargs):
+            exception_type = gcp.http_error_exception()
+
+            def try_catch_exc():
+                try:
+                    value = func(*args, **kwargs)
+                    return value
+                except Exception as e:  # pylint: disable=broad-except
+                    if not isinstance(e, exception_type) or (
+                            regex and not re.search(regex, str(e))):
+                        raise
+                    return e
+
+            for _ in range(max_retries):
+                ret = try_catch_exc()
+                if not isinstance(ret, Exception):
+                    break
+                time.sleep(retry_interval_s)
+            if isinstance(ret, Exception):
+                raise ret
+            return ret
+
+        return wrapper
+
+    return dec
+
+
+def _generate_node_name(cluster_name: str, node_suffix: str,
+                        is_head: bool) -> str:
+    """Generate node name from labels and suffix.
+
+    This is required so that the correct resource can be selected
+    when the only information autoscaler has is the name of the node.
+
+    The suffix is expected to be one of 'compute' or 'tpu'
+    (as in ``GCPNodeType``).
+    """
+    suffix_id = common_utils.base36_encode(uuid.uuid4().hex)
+    suffix = f'-{suffix_id[:INSTANCE_NAME_UUID_LEN]}-{node_suffix}'
+    if is_head:
+        suffix = f'-head{suffix}'
+    else:
+        suffix = f'-worker{suffix}'
+    node_name = cluster_name + suffix
+    assert len(node_name) <= INSTANCE_NAME_MAX_LEN, cluster_name
+    return node_name
+
+
+def _log_errors(errors: List[Dict[str, str]], e: Any, zone: str) -> None:
+    """Format errors into a string."""
+    if errors:
+        plural = 's' if len(errors) > 1 else ''
+        codes = ', '.join(repr(e.get('code', 'NA')) for e in errors)
+        messages = '; '.join(
+            repr(e.get('message', 'NA').strip('.')) for e in errors)
+        logger.warning(
+            f'create_instances: Failed with return code{plural} {codes} in '
+            f'{zone}. (message{plural}: {messages})')
```

Comment:
```suggestion
            f'create_instances: Got return code{plural} {codes} in '
            f'{zone}: {messages}')
```

Reply:
very minor, subjective: maybe get rid of the quotes (repr) around the return codes.

Reply:
I tried codes without quotes, but it looked quite confusing as some of the codes are normal words, which can be hard to distinguish from the other words in the sentence.